pyscriptjs/tests/test_py_button.py

```diff
@@ -0,0 +1,34 @@
+import pytest
+
+from .support import PyScriptTest
+
+
+class TestPyButton(PyScriptTest):
+    @pytest.mark.xfail
+    def test_on_click(self):
+        """
+        currently this test fails for a bad reason which is unrelated to
+        py-button. During the page loading, the following JS exception occur,
+        in base.ts:BaseEvalElement.evaluate
+
+        [JS exception   ] TypeError: Cannot use 'in' operator to search for 'runPythonAsync' in undefined
```

Comment:
hmm, this happens when runtime is not loaded yet, but it gets eventually loaded after.
So, should we make this test pass? by using `self.clear_errors()` perhaps?
Ideally, we should not let this occur and change so that this check doesn't happen when runtime isn't ready yet.

Reply:
It's debatable. From some point of view, I agree with you that the test "works as intended" so it should pass.
But from another point of view, the fact that we get an unexpected exception indicates that something is wrong and it should be fixed. And also, without investigating I don't really know whether this is a bug in `pybutton.ts` or somewhere else.

This is also a consequence of not having had proper tests until now. If we followed proper TDD since the beginning, we would have caught the bug as soon as it appeared. Now we are transitioning from "no tests" to (hopefully) "full test coverage", so we will uncover more and more bugs. As a team we need to decide how strict we want to be, and my personal opinion is that we should be *very* strict and not let any JS exception to happen (unless there is a good reason).

> by using self.clear_errors() perhaps?

this is a bad idea. Currently, it is clear that the test is supposed to fail, and if it does not fail `pytest` reports it as "xpass" (i.e., "unexpectedly passing").
Imagine that you add a `clear_errors()` at the end; then we fix the bug, but we forget to remove the `clear_errors()`. Weeks or months later, we introduce a bug which causes a JS exception in this test, but the test will not notice because we call `clear_errors().` Better to avoid going down that route.

> Ideally, we should not let this occur and change so that this check doesn't happen when runtime isn't ready yet.

yes, but this should be fixed in #673

Reply:
@fpliger what do you think of this here? I am also divided but have some inclination to let it pass with a comment to remove `self.clear_errors()` in future. But of course, I can be wrong.